hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/WALCellCodec.java

```diff
@@ -256,6 +278,26 @@ public void write(Cell cell) throws IOException {
         }
       }
     }
+
+    private byte[] compressValue(Cell cell) throws IOException {
+      byte[] buffer = new byte[4096];
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      Deflater deflater = compression.getValueCompressor().getDeflater();
+      deflater.setInput(cell.getValueArray(), cell.getValueOffset(), cell.getValueLength());
+      boolean finished = false;
+      do {
+        int bytesOut = deflater.deflate(buffer);
+        if (bytesOut > 0) {
+          baos.write(buffer, 0, bytesOut);
+        } else {
+          bytesOut = deflater.deflate(buffer, 0, buffer.length, Deflater.SYNC_FLUSH);
```

Comment:
If we reach this else condition that means we have compressed all the data and written to ByteArrayOutputStream. 
Now in else condition, we are sure that bytesOut will be 0. We are not resetting `buffer`  anywhere. So for the last chunk of Cell's value whose size <= 4096, will we be compressing them twice ?
@apurtell  It is entirely possible that I misunderstood something. Please correct me if that is the case. Thank you !

Reply:
Also the java doc for SYNC_FLUSH says the following thing.
```
In the case of FULL_FLUSH or SYNC_FLUSH, if the return value is len, the space available in output buffer b, this method should be invoked again with the same flush parameter and more output space.
```
https://docs.oracle.com/javase/8/docs/api/java/util/zip/Deflater.html#deflate-byte:A-int-int-int-
@apurtell  Do we need to change something regarding this comment ?

Reply:
Yes you are missing the semantics of Deflator#deflate. 

Deflator#deflate will only return 0 if it needs more input. At this point because there is no more input, we are now done, but we still need to flush. To flush it we call deflate() again using the method that allows us to specify a sync flag.

Reply:
I will add a comment here.